krm-functions/nad-fn/mutator/mutator.go

```diff
@@ -0,0 +1,190 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mutator
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	nadv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+	nephioreqv1alpha1 "github.com/nephio-project/api/nf_requirements/v1alpha1"
+	infrav1alpha1 "github.com/nephio-project/nephio-controller-poc/apis/infra/v1alpha1"
+	clusterctxtlibv1alpha1 "github.com/nephio-project/nephio/krm-functions/lib/clustercontext/v1alpha1"
+	interfacelibv1alpha1 "github.com/nephio-project/nephio/krm-functions/lib/interface/v1alpha1"
+	//ipallocv1v1alpha1 "github.com/nephio-project/nephio/krm-functions/lib/ipallocation/v1alpha1"
+	condkptsdk "github.com/nephio-project/nephio/krm-functions/lib/condkptsdk"
+	ipalloclibv1alpha1 "github.com/nephio-project/nephio/krm-functions/lib/ipallocation/v1alpha1"
+	nadlibv1 "github.com/nephio-project/nephio/krm-functions/lib/nad/v1"
+	ipamv1alpha1 "github.com/nokia/k8s-ipam/apis/alloc/ipam/v1alpha1"
+	vlanv1alpha1 "github.com/nokia/k8s-ipam/apis/alloc/vlan/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type mutatorCtx struct {
+	fnCondSdk         condkptsdk.KptCondSDK
+	masterInterface   string
+	cniType           string
+	siteCode          string
+	clusterContextSet bool
+}
+
+func Run(rl *fn.ResourceList) (bool, error) {
+	m := mutatorCtx{}
+	var err error
+	m.fnCondSdk, err = condkptsdk.New(
+		rl,
+		&condkptsdk.Config{
+			For: corev1.ObjectReference{
+				APIVersion: nadv1.SchemeGroupVersion.Identifier(),
+				Kind:       reflect.TypeOf(nadv1.NetworkAttachmentDefinition{}).Name(),
+			},
+			Watch: map[corev1.ObjectReference]condkptsdk.WatchCallbackFn{
+				{
+					APIVersion: infrav1alpha1.GroupVersion.Identifier(),
+					Kind:       reflect.TypeOf(infrav1alpha1.ClusterContext{}).Name(),
+				}: m.ClusterContextCallbackFn,
+				{
+					APIVersion: ipamv1alpha1.GroupVersion.Identifier(),
+					Kind:       ipamv1alpha1.IPAllocationKind,
+				}: nil,
+				{
+					APIVersion: vlanv1alpha1.GroupVersion.Identifier(),
+					Kind:       vlanv1alpha1.VLANAllocationKind,
+				}: nil,
+				{
+					APIVersion: nephioreqv1alpha1.GroupVersion.Identifier(),
+					Kind:       nephioreqv1alpha1.InterfaceKind,
+				}: nil,
+			},
+			PopulateOwnResourcesFn: nil,
+			GenerateResourceFn:     m.generateResourceFn,
+		},
+	)
+	if err != nil {
+		rl.Results = append(rl.Results, fn.ErrorResult(err))
+	}
+	return m.fnCondSdk.Run()
+}
+
+func (r *mutatorCtx) ClusterContextCallbackFn(o *fn.KubeObject) error {
+	clusterContext := clusterctxtlibv1alpha1.NewMutator(o.String())
```

Comment:
This seem to involve an unnecessary YAML marshal round-trip. `o` already contains a parsed version of the input YAML. 
I suggest to use `o.As()` or methods of `KubeObjectExt[ClusterContext]` instead

Reply:
I was using the ClusterContext lib which is outdated, will rework to use the validate method from API instead